### PR TITLE
The quest for a better search

### PR DESCRIPTION
**Pull request by mauroservienti**

### CODE REVIEW

File: _posts/2019-05-15-the-quest-for-a-better-search.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+As a developer if there is one thing I should blame Google for is the distortion they inflicted to the search experience for users. When customers requirements list things like “Google-like search capabilities” my comment is generally along the lines of:
+
+> Give me the Google budget, and you’ll have the Google search.
+
+Last week I was on a call with my colleague [Tomek](https://twitter.com/Masternak) to discuss search capabilities in a distributed system. The very first distinction that we made is:
+
+> You are Google, you are not Google.
+
+Most of the times the “you are not Google” mantra is used to say that if you’re not at the scale of Google something (technology or architecture choices) is not worth it. The comment is not about the scale of the business, so for example in this context Amazon is not Google: Searching is one of Google’s core businesses, that’s not the case for Amazon.
```
Comment (tmasternak) [inline _posts/2019-05-15-the-quest-for-a-better-search.md]:
``` suggestion
Most of the times the “you are not Google” mantra is used to say that if you’re not at the scale of Google something (technology or architecture choices) is not worth it. The comment is not about the scale but rather about the business domain. For example, Amazon is not Google in the sense that search is not at the center of their business.
```

File: _posts/2019-05-15-the-quest-for-a-better-search.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+How does this reflect on the architecture? In a service oriented architecture:
```
Comment (tmasternak) [inline _posts/2019-05-15-the-quest-for-a-better-search.md]:
``` suggestion
How does this reflect in the architecture? In a service-oriented architecture:
```

File: _posts/2019-05-15-the-quest-for-a-better-search.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+The simplest solution would be for all services involved in every type of supported search to:
+
+- Deliver data to a common place, for example a Sql Database or an Elasticsearch instance
+- Have the search logic backed into the aforementioned storage
+
+It works, and it’s doable. However, it violates boundaries. Is it a problem? It’s up to you to decide, it’s the typical “it depends” consultant answer.
+
+I’d say that in a small system, with a small team, with a low data change rate, and with near zero data schema evolution it’s not a problem at all. As I said: it works and brilliantly solves the problem.
+
+However it’s a huge source of coupling. As soon as the system grows, or the team gets larger, or there is higher data or data schema change rate, the hodgepodge becomes a source of pain.
```
Comment (tmasternak) [inline _posts/2019-05-15-the-quest-for-a-better-search.md]: Is this a bigger source of coupling than UI composition? 

Personally, I'm not sure - data ownership is still inside each service and the query/search model is evolved through requirements at the system level. 

I do think this is harder to evolve form each service perspective than UI composition as we are talking about persistent state - so it's yet another storage schema to maintain.

File: _posts/2019-05-15-the-quest-for-a-better-search.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+Multiple services, and thus teams, will find themselves fighting with each other to evolve document schema, to change query structure, to invalidate document sections, on so on. All these will surface also as deployment and development constraints.
```
Comment (tmasternak) [inline _posts/2019-05-15-the-quest-for-a-better-search.md]:
``` suggestion
Multiple services, and thus teams, will find themselves fighting with each other to evolve document schema, to change query structure, to invalidate document sections etc.. All these will surface also as deployment and development constraints.
```

File: _posts/2019-05-15-the-quest-for-a-better-search.md (hunk 3 above)
Comment (mauroservienti) [reply _posts/2019-05-15-the-quest-for-a-better-search.md]: Data ownership is for sure inside each service, the problem with an external storage that owns schema and query (thus data and behaviors), that is what I'm describing here, is that it results in a sort of no man's land, no one will have authority to change anything. I guess it's not dissimilar from the coupling brought in by Branding in UI Composition. It's probably worse but not dissimilar.

**Pull request merged**